File: it-agents/skills/code-architect/references/code-examples/php/Visitor.php
<?php
// Visitor — separate algorithms from objects via accept/visit (double dispatch).
declare(strict_types=1);

interface Visitor
{
    public function visitText(TextElement $element): string;
    public function visitImage(ImageElement $element): string;
}

interface Element
{
    public function accept(Visitor $visitor): string;
}

final class TextElement implements Element
{
    public function __construct(public string $content) {}

    public function accept(Visitor $visitor): string
    {
        return $visitor->visitText($this);
    }
}

final class ImageElement implements Element
{
    public function __construct(public string $src) {}

    public function accept(Visitor $visitor): string
    {
        return $visitor->visitImage($this);
    }
}

// A new operation = a new visitor, no element changes needed.
final class HtmlExportVisitor implements Visitor
{
    public function visitText(TextElement $element): string
    {
        return "<p>{$element->content}</p>";
    }

    public function visitImage(ImageElement $element): string
    {
        return "<img src=\"{$element->src}\">";
    }
}

final class PlainTextVisitor implements Visitor
{
    public function visitText(TextElement $element): string { return $element->content; }
    public function visitImage(ImageElement $element): string { return "[image: {$element->src}]"; }
}

// Demo
$elements = [new TextElement('Hello'), new ImageElement('logo.png')];
foreach ([new HtmlExportVisitor(), new PlainTextVisitor()] as $visitor) {
    foreach ($elements as $element) {
        echo $element->accept($visitor) . PHP_EOL;
    }
}

File: it-agents/skills/code-architect/references/code-examples/csharp/Visitor.cs
// Visitor — separate algorithms from the objects they operate on via accept/visit (double dispatch).
using System;
using System.Collections.Generic;

namespace Patterns.Visitor
{
    interface IVisitor
    {
        string Visit(ElementA element);
        string Visit(ElementB element);
    }

    abstract class Element
    {
        public abstract string Accept(IVisitor visitor);
    }

    sealed class ElementA : Element
    {
        public int Number { get; } = 7;
        public override string Accept(IVisitor visitor) => visitor.Visit(this);
    }

    sealed class ElementB : Element
    {
        public string Text { get; } = "node";
        public override string Accept(IVisitor visitor) => visitor.Visit(this);
    }

    // A concrete operation defined entirely outside the element classes.
    sealed class DescribeVisitor : IVisitor
    {
        public string Visit(ElementA element) => $"A holds number {element.Number}";
        public string Visit(ElementB element) => $"B holds text '{element.Text}'";
    }

    static class Demo
    {
        public static void Run()
        {
            var elements = new List<Element> { new ElementA(), new ElementB() };
            var visitor = new DescribeVisitor();
            foreach (Element element in elements)
                Console.WriteLine(element.Accept(visitor));
        }

        public static void Main() => Run();
    }
}
